File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java
package frc.robot.subsystems.shooter;

import org.littletonrobotics.junction.AutoLog;

public interface ShooterIO {
  @AutoLog
  public static class ShooterIOInputs {
    public double lowerPositionDegrees = 0.0;
    public double lowerVelocityRpm = 0.0;
    public double lowerAppliedVolts = 0.0;
    public double lowerRpmSetpoint = 0.0;
    public double lowerSurfaceSpeed = 0.0;
    public double lowerOutput = 0.0;
    public double[] lowerCurrentAmps = new double[] {};

    public double upperPositionDegrees = 0.0;
    public double upperVelocityRpm = 0.0;
    public double upperAppliedVolts = 0.0;
    public double upperRpmSetpoint = 0.0;
    public double upperSurfaceSpeed = 0.0;
    public double upperOutput = 0.0;
    public double[] upperCurrentAmps = new double[] {};
  }

  /** Updates the set of loggable inputs. */
  public default void updateInputs(ShooterIOInputs inputs) {}

  /** Run open loop at the specified voltage. */
  public default void set(double lowerPercent, double upperPercent) {}

  /** Run closed loop at the specified velocity. */
  public default void setVelocity(
      double lowerRpm, double upperRpm) {}

  public default void setVoltage(double volts) {}


}

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.kt
package frc.robot.subsystems.shooter

import org.littletonrobotics.junction.AutoLog

/**
 *
 */
interface ShooterIO {
    /**
     *
     */
    @AutoLog
    open class ShooterIOInputs {
        /**
         *
         */
        @JvmField
        var lowerPositionDegrees: Double = 0.0
        /**
         *
         */
        @JvmField
        var lowerVelocityRpm: Double = 0.0
        /**
         *
         */
        @JvmField
        var lowerAppliedVolts: Double = 0.0
        /**
         *
         */
        @JvmField
        var lowerRpmSetpoint: Double = 0.0
        /**
         *
         */
        @JvmField
        var lowerSurfaceSpeed: Double = 0.0
        /**
         *
         */
        @JvmField
        var lowerOutput: Double = 0.0
        /**
         *
         */
        @JvmField
        var lowerCurrentAmps: DoubleArray = doubleArrayOf()

        /**
         *
         */
        @JvmField
        var upperPositionDegrees: Double = 0.0
        /**
         *
         */
        @JvmField
        var upperVelocityRpm: Double = 0.0
        /**
         *
         */
        @JvmField
        var upperAppliedVolts: Double = 0.0
        /**
         *
         */
        @JvmField
        var upperRpmSetpoint: Double = 0.0
        /**
         *
         */
        @JvmField
        var upperSurfaceSpeed: Double = 0.0
        /**
         *
         */
        @JvmField
        var upperOutput: Double = 0.0
        /**
         *
         */
        @JvmField
        var upperCurrentAmps: DoubleArray = doubleArrayOf()
    }

    /** Updates the set of loggable inputs.  */
    fun updateInputs(inputs: ShooterIOInputs) {}

    /** Run open loop at the specified voltage.  */
    operator fun set(lowerPercent: Double, upperPercent: Double) {}

    /** Run closed loop at the specified velocity.  */
    fun setVelocity(
        lowerRpm: Double, upperRpm: Double
    ) {
    }

    /**
     *
     */
    fun setVoltage(volts: Double) {}
}
